h5netcdf/core.py

```diff
@@ -150,15 +150,10 @@ def __array__(self, *args, **kwargs):
         return self._h5ds.__array__(*args, **kwargs)
 
     def __getitem__(self, key):
-        if h5py.__version__ >= LooseVersion("3.0.0") and self._root.decode_strings:
-            if self.dtype == str:
+        if getattr(self._root, "decode_strings", False):
+            # need to wrap with np.array to fetch .dtype in any case
+            if h5py.check_string_dtype(np.array(self._h5ds[key]).dtype) is not None:
```

Comment:
This is potentially problematic, because converting into a numpy array loads the entire HDF5 array into memory.

Can you give an example of why this fails?

Reply:
Ah, didn't think about that. The problem is that 'bytes' do not have dtype and then it breaks. That was the reason for the special checks for 'bytes'.

Reply:
So if `self._h5ds[key]` is of type `bytes`, then `self. _h5ds[key].dtype` raises `AttributeError`.

Reply:
> So if `self._h5ds[key]` is of type `bytes`

When can this happen? What does the underlying HDF5 file look like? As far as I can tell, that shouldn't happen...

Reply:
The problem is with the variable length string. 

Simple example to reproduce:

```python
    ds = netCDF4.Dataset(tmp_local_netcdf, "w")
    ds.createDimension("x", 4)
    v = ds.createVariable("var_len_str", str, ("x"))
    v[0] = u"foo"
    ds.close()

    ds = h5netcdf.legacyapi.Dataset(tmp_local_netcdf, "r")
    v = ds.variables["var_len_str"]
    assert v.dtype == str
    assert v[0] == u"foo"
```

Error traceback:
```
<h5netcdf.legacyapi.Variable '/var_len_str': dimensions ('x',), shape (4,), dtype <class 'str'>>
Attributes:
<HDF5 dataset "var_len_str": shape (4,), type "|O">
b'foo'

h5netcdf/tests/test_h5netcdf.py:120 (test_bytes)
tmp_local_netcdf = '/tmp/pytest-of-kai/pytest-8/test_bytes0/testfile.nc'

    def test_bytes(tmp_local_netcdf):
        ds = netCDF4.Dataset(tmp_local_netcdf, "w")
        ds.createDimension("x", 4)
        v = ds.createVariable("var_len_str", str, ("x"))
        v[0] = u"foo"
        ds.close()
    
        ds = h5netcdf.legacyapi.Dataset(tmp_local_netcdf, "r")
        v = ds.variables["var_len_str"]
        assert v.dtype == str
>       assert v[0] == u"foo"

h5netcdf/tests/test_h5netcdf.py:131: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <h5netcdf.legacyapi.Variable '/var_len_str': dimensions ('x',), shape (4,), dtype <class 'str'>>
Attributes:
key = 0

    def __getitem__(self, key):
        if getattr(self._root, "decode_strings", False):
            # need to wrap with np.array to fetch .dtype in any case
            try:
>               if h5py.check_string_dtype(self._h5ds[key].dtype) is not None:
E               AttributeError: 'bytes' object has no attribute 'dtype'

h5netcdf/core.py:156: AttributeError

```
h5dump of testfile.nc

```
(h5py3) kai@linux:~$ h5dump /tmp/pytest-of-kai/pytest-7/test_bytes0/testfile.nc
HDF5 "/tmp/pytest-of-kai/pytest-7/test_bytes0/testfile.nc" {
GROUP "/" {
   ATTRIBUTE "_NCProperties" {
      DATATYPE  H5T_STRING {
         STRSIZE 34;
         STRPAD H5T_STR_NULLTERM;
         CSET H5T_CSET_ASCII;
         CTYPE H5T_C_S1;
      }
      DATASPACE  SCALAR
      DATA {
      (0): "version=2,netcdf=4.7.4,hdf5=1.10.6"
      }
   }
   DATASET "var_len_str" {
      DATATYPE  H5T_STRING {
         STRSIZE H5T_VARIABLE;
         STRPAD H5T_STR_NULLTERM;
         CSET H5T_CSET_UTF8;
         CTYPE H5T_C_S1;
      }
      DATASPACE  SIMPLE { ( 4 ) / ( 4 ) }
      DATA {
      (0): "foo", "", "", ""
      }
      ATTRIBUTE "DIMENSION_LIST" {
         DATATYPE  H5T_VLEN { H5T_REFERENCE { H5T_STD_REF_OBJECT }}
         DATASPACE  SIMPLE { ( 1 ) / ( 1 ) }
         DATA {
         (0): (DATASET 331 /x )
         }
      }
   }
   DATASET "x" {
      DATATYPE  H5T_IEEE_F32BE
      DATASPACE  SIMPLE { ( 4 ) / ( 4 ) }
      DATA {
      (0): 0, 0, 0, 0
      }
      ATTRIBUTE "CLASS" {
         DATATYPE  H5T_STRING {
            STRSIZE 16;
            STRPAD H5T_STR_NULLTERM;
            CSET H5T_CSET_ASCII;
            CTYPE H5T_C_S1;
         }
         DATASPACE  SCALAR
         DATA {
         (0): "DIMENSION_SCALE"
         }
      }
      ATTRIBUTE "NAME" {
         DATATYPE  H5T_STRING {
            STRSIZE 64;
            STRPAD H5T_STR_NULLTERM;
            CSET H5T_CSET_ASCII;
            CTYPE H5T_C_S1;
         }
         DATASPACE  SCALAR
         DATA {
         (0): "This is a netCDF dimension but not a netCDF variable.         4"
         }
      }
      ATTRIBUTE "REFERENCE_LIST" {
         DATATYPE  H5T_COMPOUND {
            H5T_REFERENCE { H5T_STD_REF_OBJECT } "dataset";
            H5T_STD_I32LE "dimension";
         }
         DATASPACE  SIMPLE { ( 1 ) / ( 1 ) }
         DATA {
         (0): {
               DATASET 685 /var_len_str ,
               0
            }
         }
      }
      ATTRIBUTE "_Netcdf4Dimid" {
         DATATYPE  H5T_STD_I32LE
         DATASPACE  SCALAR
         DATA {
         (0): 0
         }
      }
   }
}
}
```

Reply:
Good chance there is some quirk in my logic.

Reply:
@shoyer The latest change catches the `AttributeError` and assumes that these `bytes` arrays should be treated with `.asstr()` too. This prevents the unconditional loading via the former `np.array()`-wrapper. I'm really not sure, there might be a better solution, though.

Reply:
@shoyer  Now, after thinking a bit more about that, we should test against the dtype of the dataset, which works quite nice.

Reply:
Fixed by checking against `self._h5ds.dtype` instead of `self._h5ds[key].dtype`.

I've looked at these two lines of the traceback for quite some time until I got it:

```
        assert v.dtype == str
>       assert v[0] == u"foo"
```

Reply:
Ah, excellent! Yes, that's exactly what I was trying to suggest